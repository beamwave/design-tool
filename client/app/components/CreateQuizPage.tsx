import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import fontawesome from '@fortawesome/fontawesome'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { startCreateNewQuiz } from '../actions/create'

interface IProps {
  history?: any
  username?: any
  quizzes?: any
  startCreateNewQuiz: (quiz: any, any) => any
}

export class CreateQuizPage extends Component<IProps, any> {
  state = {
    quiz: {
      author: this.props.username,
      title: '',
      description: '',
      isPrivate: true,
      type: 'quiz'
    },
    quizzes: this.props.quizzes
  }

  private updateQuiz = (e: any, arg1: string) => {
    switch (arg1) {
      case 'title':
        this.setState({
          quiz: {
            ...this.state.quiz,
            title: e.target.value
          }
        })
        break
      case 'description':
        this.setState({
          quiz: {
            ...this.state.quiz,
            description: e.target.value
          }
        })
      default:
        break
    }
    // console.log(this.state)
  }

  private updatePrivate = (e: any) => {
    switch (e.target.value) {
      case 'private':
        this.setState({
          quiz: {
            ...this.state.quiz,
            isPrivate: true
          }
        })
        break

      default:
        this.setState({
          quiz: {
            ...this.state.quiz,
            isPrivate: false
          }
        })
        break
    }
  }

  private submitQuiz = async (e: any) => {
    e.preventDefault()
    const { startCreateNewQuiz, username } = this.props
    const { quiz } = this.state

    // save quiz to db and update quizzes
    const createAction = await startCreateNewQuiz(quiz, this.props.history)

    // this.props.history.push('/quizzes/created')
    // this.props.history.push('/edit-quiz/' + uuid)
    // .then(res => {
    //   console.log('add quiz props')
    //   console.log(this.props)
    //   this.props.history.push('/add-question')
    // })
    // .catch(err =>
    //   console.log(
    //     'make sure the user knows that the quiz did not make it through'
    //   )
    // )
  }

  // @ts-ignore
  render = () => {
    const {
      quiz: { title }
    } = this.state
    return (
      <div className="create-quiz-page">
        <main>
          <div
            className={
              title.length > 0
                ? 'create-quiz-container valid'
                : 'create-quiz-container invalid'
            }
          >
            <p className="title">Create a quiz</p>
            <form className="details">
              <div className="group">
                <label htmlFor="create-quiz-title">Title</label>
                <input
                  type="text"
                  id="create-quiz-title"
                  value={this.state.quiz.title}
                  onChange={(e: any) => {
                    this.updateQuiz(e, 'title')
                  }}
                />
              </div>
              <div className="group">
                <label htmlFor="create-quiz-description">Description</label>
                <textarea
                  id="create-quiz-description"
                  value={this.state.quiz.description}
                  rows={3}
                  onChange={(e: any) => {
                    this.updateQuiz(e, 'description')
                  }}
                />
              </div>
              <div className="group">
                <label htmlFor="owner">Owner</label>
                <FontAwesomeIcon className="icon" icon="angle-down" />
                <select
                  name="owner"
                  id="question-options-dropdown"
                  onChange={this.updatePrivate}
                >
                  <option value="choose" hidden>
                    select
                  </option>
                  <option value="private">private</option>
                  <option value="public">public</option>
                </select>
              </div>
              <button
                type="button"
                className="save-quiz"
                onClick={this.submitQuiz}
                disabled={title.length > 0 ? false : true}
              >
                Create
              </button>
            </form>
          </div>
        </main>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  username: state.auth.username,
  quizzes: state.quizzes.all
})

export default connect<any, IProps>(
  mapStateToProps,
  { startCreateNewQuiz }
)(CreateQuizPage)
