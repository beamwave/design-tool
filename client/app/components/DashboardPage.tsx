import React, { Component } from 'react'
import { connect } from 'react-redux'
import ReactGridGallery from 'react-grid-gallery'
import { FontAwesomeIcon } from '../../../node_modules/@fortawesome/react-fontawesome'
import { check } from '../helpers/helpers'
import { filter } from '../actions/filter'

interface FData {
  category: string
  value: string
}

interface IProps {
  username: string
  category: string
  categories?: any
  images: any[]
  filters?: any[]
  filter?: (by: FData) => any
  history?: any
}

interface IState {
  images: any[]
}

export class DashboardPage extends Component<IProps, IState> {
  state = {
    images: [],
    categories: 'map an object'
  }

  // filter images
  handleFilterChange = ({ target }) => {
    const { filters, filter } = this.props
    // "medium", "device", "wardrobe", "mood", etc.
    // console.log('dataset', target.dataset.filterName)
    // console.log('value', target.value)

    const by = {
      category: target.dataset.filterName,
      value: target.value
    }

    filter(by)
  }

  handleThumbnailClick = (i, e) => {
    const { images, history } = this.props
    const imageId = images[i]._id
    history.push(`/image/${imageId}`)
  }

  componentDidMount = () => {
    const { images: imagesFromServer } = this.props
    console.log('imagesFromServer', imagesFromServer)
    let images
    if (check(imagesFromServer) && Array.isArray(imagesFromServer))
      images = imagesFromServer.map(image => ({
        src: image.url,
        thumbnail: image.url,
        thumbnailHeight: image.height,
        thumbnailWidth: image.width
      }))

    this.setState({ images })
  }

  render = () => {
    const {
      category,
      categories,
      images: imagesFromServer,
      filters
    } = this.props
    const { images } = this.state
    const alive = check(imagesFromServer)
    return (
      <main className="dashboard-page">
        {
          <div>
            <section className="image-results">
              {alive &&
                imagesFromServer.length === 0 && (
                  <p className="null-case">No images to display.</p>
                )}
              {alive &&
                imagesFromServer.length > 0 && (
                  <ReactGridGallery
                    images={check(imagesFromServer) ? images : []}
                    enableImageSelection={false}
                    onClickThumbnail={this.handleThumbnailClick}
                  />
                )}
            </section>
            <aside className="primary-config">
              {category === 'art' && (
                <main>
                  <form className="filter-container">
                    <label htmlFor="select">medium</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['medium']}
                      data-filter-name="medium"
                      onChange={this.handleFilterChange}
                    >
                      {categories !== undefined &&
                      categories.hasOwnProperty('medium') ? (
                        categories['medium'].map(option => (
                          <option key={option.name} value={option.name}>
                            {option.name}
                          </option>
                        ))
                      ) : (
                        <option value="n/a">n/a</option>
                      )}
                    </select>
                  </form>
                  <form className="filter-container">
                    <label htmlFor="select">device</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['device']}
                      data-filter-name="device"
                      onChange={this.handleFilterChange}
                    >
                      {categories !== undefined &&
                      categories.hasOwnProperty('device') ? (
                        categories['medium'].map(option => (
                          <option key={option.name} value={option.name}>
                            {option.name}
                          </option>
                        ))
                      ) : (
                        <option value="n/a">n/a</option>
                      )}
                    </select>
                  </form>
                  <form className="filter-container">
                    <label htmlFor="select">project</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['project']}
                      data-filter-name="project"
                      onChange={this.handleFilterChange}
                    >
                      {categories !== undefined &&
                      categories.hasOwnProperty('device') ? (
                        categories['medium'].map(option => (
                          <option key={option.name} value={option.name}>
                            {option.name}
                          </option>
                        ))
                      ) : (
                        <option value="n/a">n/a</option>
                      )}
                    </select>
                  </form>
                </main>
              )}
              {category === 'comments' && (
                <main>
                  <form className="filter-container">
                    <label htmlFor="select">topic</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['topic']}
                      data-filter-name="topic"
                      onChange={this.handleFilterChange}
                    >
                      <option value="art">art</option>
                      <option value="comments">comments</option>
                      <option value="clothes">clothes</option>
                      <option value="people">people</option>
                    </select>
                  </form>
                  <form className="filter-container">
                    <label htmlFor="select">style</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['style']}
                      data-filter-name="style"
                      onChange={this.handleFilterChange}
                    >
                      <option value="art">art</option>
                      <option value="comments">comments</option>
                      <option value="clothes">clothes</option>
                      <option value="people">people</option>
                    </select>
                  </form>
                  <form className="filter-container">
                    <label htmlFor="select">site</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['site']}
                      data-filter-name="site"
                      onChange={this.handleFilterChange}
                    >
                      <option value="art">art</option>
                      <option value="comments">comments</option>
                      <option value="clothes">clothes</option>
                      <option value="people">people</option>
                    </select>
                  </form>
                </main>
              )}
              {category === 'clothes' && (
                <main>
                  <form className="filter-container">
                    <label htmlFor="select">wardrobe</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['wardrobe']}
                      data-filter-name="wardrobe"
                      onChange={this.handleFilterChange}
                    >
                      <option value="art">art</option>
                      <option value="comments">comments</option>
                      <option value="clothes">clothes</option>
                      <option value="people">people</option>
                    </select>
                  </form>
                  <form className="filter-container">
                    <label htmlFor="select">attire</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['attire']}
                      data-filter-name="attire"
                      onChange={this.handleFilterChange}
                    >
                      <option value="art">art</option>
                      <option value="comments">comments</option>
                      <option value="clothes">clothes</option>
                      <option value="people">people</option>
                    </select>
                  </form>
                </main>
              )}
              {category === 'people' && (
                <main>
                  <form className="filter-container">
                    <label htmlFor="select">mood</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['mood']}
                      data-filter-name="mood"
                      onChange={this.handleFilterChange}
                    >
                      <option value="art">art</option>
                      <option value="comments">comments</option>
                      <option value="clothes">clothes</option>
                      <option value="people">people</option>
                    </select>
                  </form>
                  <form className="filter-container">
                    <label htmlFor="select">group</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['group']}
                      data-filter-name="group"
                      onChange={this.handleFilterChange}
                    >
                      <option value="art">art</option>
                      <option value="comments">comments</option>
                      <option value="clothes">clothes</option>
                      <option value="people">people</option>
                    </select>
                  </form>
                  <form className="filter-container">
                    <label htmlFor="select">angle</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['angle']}
                      data-filter-name="angle"
                      onChange={this.handleFilterChange}
                    >
                      <option value="art">art</option>
                      <option value="comments">comments</option>
                      <option value="clothes">clothes</option>
                      <option value="people">people</option>
                    </select>
                  </form>
                </main>
              )}
            </aside>
            <aside className="secondary-config">
              {category === 'art' && (
                <main>
                  <div className="filter-group">
                    <div className="input-group">
                      <label htmlFor="color">color</label>
                      <input
                        className="input"
                        name="color"
                        type="text"
                        placeholder="#hex or name"
                      />
                      <div className="color-group">
                        <div className="color-1" />
                        <div className="color-2" />
                        <div className="color-3" />
                        <FontAwesomeIcon icon="angle-right" className="icon" />
                      </div>
                    </div>
                  </div>
                  <div className="filter-group">
                    <div className="input-group">
                      <label htmlFor="length">length</label>
                      <input name="length" type="text" placeholder="px" />
                    </div>
                    <div className="input-group">
                      <label htmlFor="width">width</label>
                      <input name="width" type="text" placeholder="px" />
                    </div>
                  </div>
                </main>
              )}
              {category === 'comments' && (
                <main>
                  <div className="filter-group">
                    <div className="input-group">
                      <label htmlFor="username">username</label>
                      <input className="names" name="username" type="text" />
                    </div>
                  </div>
                </main>
              )}
              {category === 'clothes' && (
                <main>
                  <div className="filter-group">
                    <div className="input-group">
                      <label htmlFor="color">color</label>
                      <input
                        className="input"
                        name="color"
                        type="text"
                        placeholder="#hex or name"
                      />
                      <div className="color-group">
                        <div className="color-1" />
                        <div className="color-2" />
                        <div className="color-3" />
                        <FontAwesomeIcon icon="angle-right" className="icon" />
                      </div>
                    </div>
                  </div>
                  <div className="filter-group gender">
                    <div className="input-group">
                      <label>gender</label>
                      <FontAwesomeIcon icon="male" className="male-icon" />
                    </div>
                    <div className="input-group">
                      <FontAwesomeIcon icon="female" className="female-icon" />
                    </div>
                  </div>
                </main>
              )}
              {category === 'people' && (
                <main>
                  <div className="filter-group">
                    <div className="input-group">
                      <label htmlFor="name">name</label>
                      <input className="names" name="name" type="text" />
                    </div>
                  </div>
                  <div className="filter-group gender">
                    <div className="input-group">
                      <label>gender</label>
                      <FontAwesomeIcon icon="male" className="male-icon" />
                    </div>
                    <div className="input-group">
                      <FontAwesomeIcon icon="female" className="female-icon" />
                    </div>
                  </div>
                </main>
              )}
            </aside>
          </div>
        }
      </main>
    )
  }
}

const mapStateToProps = (state, props) => ({
  token: state.auth.token,
  username: state.auth.username,
  category: state.app.category,
  categories: state.images.categories,
  images: state.images.all,
  filters: state.filters
})

export default connect<any>(
  mapStateToProps,
  { filter }
)(DashboardPage)
