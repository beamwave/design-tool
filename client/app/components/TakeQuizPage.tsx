import React, { Component } from 'react'
import moment from 'moment'
import numeral from 'numeral'
import ImageGallery from 'react-grid-gallery'
import { connect } from 'react-redux'
import { RouteProps } from 'react-router'
import { Link } from 'react-router-dom'
import {
  startAddAnswerToArray,
  changeQuestionNumber,
  addMultipleSelectAnswer,
  addMultipleChoiceAnswer,
  updateMultipleSelectAnswer,
  updateAnswerArray,
  submitQuizAttempt
} from '../actions/quizzes'
import { loadModal } from '../actions/modal'
import {
  SUBMIT_QUIZ_MODAL,
  REPORT_QUESTION_MODAL
} from '../constants/modaltypes'
import Spinner from 'react-spinkit'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import {
  Dropdown,
  DropdownToggle,
  DropdownMenu,
  DropdownItem
} from 'reactstrap'

interface IProps extends RouteProps {
  username: any
  quiz: any
  questionNumber: any
  multipleSelectAnswer: any
  multipleChoiceAnswer: any
  answerArray: any
  history?: any
  loadModal: (type: string, title?: string, uuid?: string) => void
  changeQuestionNumber: (questionNumber: number) => void
  startAddAnswerToArray: (answerObj: {}) => void
  addMultipleChoiceAnswer: (answerObj: {}) => void
  addMultipleSelectAnswer: (answerObj: {}) => void
  updateMultipleSelectAnswer: (answerArray: any) => void
  updateAnswerArray: (answerArray: any[]) => void
  submitQuizAttempt: (
    quizUUID: string,
    user: string,
    attemptUUID: string,
    answerArray: any[]
  ) => void
}

export class TakeQuizPage extends Component<IProps, any> {
  constructor(props) {
    super(props)
  }
  state = {
    dropdownOpen: false
  }

  params = window.location.href.split('/')
  quizUUID = this.params[4]

  toggleDropdown = () =>
    this.setState((prevState: any) => ({
      dropdownOpen: !prevState.dropdownOpen
    }))

  submitQuizModal = () => this.props.loadModal(SUBMIT_QUIZ_MODAL)

  public reportQuestionModal = questionUUID => e => {
    console.log('reporting question', questionUUID)
    this.props.loadModal(REPORT_QUESTION_MODAL, null, questionUUID)
  }

  // will take you to the previous question and update answer array
  // if a choice is selected
  public previousQuizQuestion = (choices: object) => (e: any) => {
    e.preventDefault()
    if (
      this.props.multipleSelectAnswer.answer.length > 0 ||
      this.props.multipleChoiceAnswer !== null
    ) {
      if (
        choices[this.props.questionNumber].format === 'multiple-choice' ||
        choices[this.props.questionNumber].format === 'true-false'
      ) {
        this.props.startAddAnswerToArray(this.props.multipleChoiceAnswer)
        this.props.changeQuestionNumber(this.props.questionNumber - 1)
      } else if (
        choices[this.props.questionNumber].format === 'multiple-select'
      ) {
        this.props.startAddAnswerToArray(this.props.multipleSelectAnswer)
        this.props.changeQuestionNumber(this.props.questionNumber - 1)
      }
    } else {
      this.props.changeQuestionNumber(this.props.questionNumber - 1)
    }
  }
  // will take you to the next question and update answer array
  // if a choice is selected
  public nextQuizQuestion = (choices: object) => (e: any) => {
    e.preventDefault()
    console.log('here on submit', choices[this.props.questionNumber].format)
    if (
      this.props.multipleSelectAnswer.answer.length > 0 ||
      this.props.multipleChoiceAnswer !== null
    ) {
      if (
        choices[this.props.questionNumber].format === 'multiple-choice' ||
        choices[this.props.questionNumber].format === 'true-false'
      ) {
        this.props.startAddAnswerToArray(this.props.multipleChoiceAnswer)
        this.props.changeQuestionNumber(this.props.questionNumber + 1)
      } else if (
        choices[this.props.questionNumber].format === 'multiple-select'
      ) {
        this.props.startAddAnswerToArray(this.props.multipleSelectAnswer)
        this.props.changeQuestionNumber(this.props.questionNumber + 1)
      }
    } else {
      this.props.changeQuestionNumber(this.props.questionNumber + 1)
    }
  }

  // adds an answer to the answer array if the question has not already
  // been answered if it has the answer will be updated
  public addAnswerToObject = (choices: object, answer: any) => {
    let index
    let newArrayAnswers
    let newArray
    switch (choices[this.props.questionNumber].format) {
      case 'multiple-choice':
        console.log(this.props.answerArray)
        if (
          this.props.answerArray.some((obj, i) => {
            if (obj.title === choices[this.props.questionNumber].title) {
              console.log('checking answerArray')
              console.log(obj.title)
              index = i
              return true
            }
          })
        ) {
          let newArray = this.props.answerArray
          newArray.splice(index, 1)
          console.log(newArray)
          updateAnswerArray({
            newArray
          })
        }
        this.props.addMultipleChoiceAnswer({
          author: choices[this.props.questionNumber].author,
          title: choices[this.props.questionNumber].title,
          answer: answer.answer.answer,
          selected: true
        })
        break
      case 'multiple-select':
        if (
          this.props.answerArray.some((obj, i) => {
            if (obj.title === choices[this.props.questionNumber].title) {
              console.log('checking answerArray')
              console.log(obj.title)
              index = i
              return true
            }
          })
        ) {
          newArrayAnswers = this.props.answerArray[index].answer
          this.props.addMultipleSelectAnswer({
            author: choices[this.props.questionNumber].author,
            title: choices[this.props.questionNumber].title,
            answer: newArrayAnswers
          })
          console.log(newArrayAnswers)
          // CLEARS ANSWERARRAY
          let newArray = this.props.answerArray
          newArray.splice(index, 1)
          updateAnswerArray({
            newArray
          })
        }

        if (
          this.props.multipleSelectAnswer.answer.includes(
            answer.answer.answer
          ) ||
          (newArrayAnswers !== undefined &&
            newArrayAnswers.includes(answer.answer.answer))
        ) {
          console.log('found a match, now remove it', answer.answer.answer)
          if (newArrayAnswers !== undefined) {
            const index = newArrayAnswers.indexOf(answer.answer.answer)
            newArray = newArrayAnswers
            newArray.splice(index, 1)
            this.props.addMultipleSelectAnswer({
              author: choices[this.props.questionNumber].author,
              title: choices[this.props.questionNumber].title,
              answer: newArray
            })
          } else {
            const index = this.props.multipleSelectAnswer.answer.indexOf(
              answer.answer.answer
            )
            newArray = this.props.multipleSelectAnswer.answer
            console.log(newArray)
            newArray.splice(index, 1)
            this.props.addMultipleSelectAnswer({
              author: choices[this.props.questionNumber].author,
              title: choices[this.props.questionNumber].title,
              answer: newArray
            })
          }
        } else {
          newArray = this.props.multipleSelectAnswer.answer
          console.log(newArray)
          newArrayAnswers !== undefined &&
            newArrayAnswers.push(answer.answer.answer)
          console.log(newArrayAnswers)
          newArray.push(answer.answer.answer)
          this.props.addMultipleSelectAnswer({
            author: choices[this.props.questionNumber].author,
            title: choices[this.props.questionNumber].title,
            answer: newArrayAnswers !== undefined ? newArrayAnswers : newArray
          })
        }
        break

      case 'true-false':
        console.log(this.props.answerArray)
        if (
          this.props.answerArray.some((obj, i) => {
            if (obj.title === choices[this.props.questionNumber].title) {
              console.log('checking answerArray')
              console.log(obj.title)
              index = i
              return true
            }
          })
        ) {
          let newArray = this.props.answerArray
          newArray.splice(index, 1)
          console.log(newArray)
          updateAnswerArray({
            newArray
          })
        }
        this.props.addMultipleChoiceAnswer({
          author: choices[this.props.questionNumber].author,
          title: choices[this.props.questionNumber].title,
          answer: answer.answer.answer,
          selected: true
        })
        break
    }
  }

  public submit = (choices: object) => (e: any) => {
    console.log('in submit function')
    if (
      this.props.multipleSelectAnswer.answer.length > 0 ||
      this.props.multipleChoiceAnswer !== null
    ) {
      if (
        choices[this.props.questionNumber].format === 'multiple-choice' ||
        choices[this.props.questionNumber].format === 'true-false'
      ) {
        this.props.startAddAnswerToArray(this.props.multipleChoiceAnswer)
      } else if (
        choices[this.props.questionNumber].format === 'multiple-select'
      ) {
        this.props.startAddAnswerToArray(this.props.multipleSelectAnswer)
      }
    }
    this.submitQuizModal()
  }

  handleSelection = (choice, title) => {
    const {
      multipleChoiceAnswer: multCh,
      multipleSelectAnswer: multSe,
      answerArray: ansArr,
      questionNumber: curr,
      quiz
    } = this.props

    const quizTitles = quiz.questions.map(question => question.title)

    /*birth check*/
    const mca = multCh !== null && multCh !== undefined

    const msa =
      multCh === null && multSe !== undefined && multSe.answer.length > 0

    /* check previous answers */
    const aarr =
      !mca &&
      multSe !== undefined &&
      multSe.answer.length === 0 &&
      ansArr.length > 0 &&
      ansArr.some(ans => quizTitles.includes(title))
    // ansArr.length >= curr
    // ansArr[curr] !== undefined

    /* display defaults */
    const choose =
      !mca &&
      multSe !== undefined &&
      multSe.answer.length === 0 &&
      ansArr.every(ans => !quizTitles.includes(title))

    // console.log('choice', choice)
    // console.log('title', title)
    // console.log('quiz titles', quizTitles)

    // console.log(
    //   'quiz titles does include the title',
    //   ansArr.some(ans => quizTitles.includes(title))
    // )

    // console.log(
    //   'quiz titles does not include the title',
    //   ansArr.every(ans => !quizTitles.includes(title))
    // )
    // ansArr.length < curr
    // ansArr[curr] === undefined
    // const aarr =
    //   multCh === null &&
    //   (multSe.answer.length === 0 || multSe !== undefined) &&
    //   ansArr.length > 0

    // console.log('msa', msa)
    // console.log('multSe', multSe)

    /*answer comparison*/

    // check multiple choice object
    if (mca) {
      console.log('checking multiple choice...')
      return multCh.answer === choice ? 'choice selected' : 'choice'
    }

    // check multiple select object
    if (msa) {
      // if (msa && multSe.answers !== undefined) {
      console.log(
        'checking multiple select...',
        multSe.answer.some(answer => answer === choice)
      )
      return multSe.answer.some(answer => answer === choice)
        ? multSe.answer
            .map(answer => {
              if (answer === choice) return 'choice selected'
            })
            .join('')
        : 'choice'
    }

    // check answer array
    if (aarr) {
      // console.log('here is answers array', ansArr)
      const arr = ansArr.map(answer => {
        // let finalClass = ansArr.map(answer => {
        console.log('logging tritle? ', answer.title)
        let className = ''
        if (!Array.isArray(answer.answer)) {
          if (answer.answer === choice && answer.title === title) {
            className = 'choice selected'
            console.log('not an array, className: ', className)
          }
          if (answer.answer !== choice) {
            className = 'choice'
            console.log('not an array, className: ', className)
          }
        }
        if (Array.isArray(answer.answer) && answer.title === title) {
          answer.answer.includes(choice)
            ? (className = 'choice selected')
            : (className = 'choice')
          console.log('is an array, className is now: ', className)
        }
        console.log('before submission, className is', className)
        return className
      })

      return arr.includes('choice selected') ? 'choice selected' : 'choice'
    }

    if (choose) {
      return 'choice'
    }
  }

  // @ts-ignore
  render = () => {
    const {
      quiz,
      questionNumber,
      multipleChoiceAnswer,
      multipleSelectAnswer,
      answerArray
    } = this.props
    return (
      <div className="take-quiz-page">
        {quiz.questions === undefined && (
          <Spinner className="loading-indicator" name="ball-spin-fade-loader" />
        )}
        {quiz.questions !== undefined && (
          <div className="main">
            <header>
              <div className="meta">
                <div className="container">
                  <p className="current">Question {questionNumber + 1}</p>
                  <div className="icon">
                    <Dropdown
                      isOpen={this.state.dropdownOpen}
                      toggle={this.toggleDropdown}
                      className="app-dropdown-root"
                    >
                      <DropdownToggle className="dropdown-toggle">
                        <FontAwesomeIcon icon="ellipsis-h" className="menu" />
                      </DropdownToggle>
                      <DropdownMenu
                        left="true"
                        className="dropdown-menu"
                        style={{
                          display: this.state.dropdownOpen ? 'block' : 'none'
                        }}
                      >
                        <DropdownItem
                          className="dropdown-item"
                          onClick={this.reportQuestionModal(
                            this.props.quiz.questions[questionNumber].uuid
                          )}
                        >
                          Report question
                        </DropdownItem>
                      </DropdownMenu>
                    </Dropdown>
                  </div>
                </div>
                <p className="title">{quiz.title}</p>
                <p className="progress">
                  Question {questionNumber + 1}/{quiz.questions.length}
                </p>
                <p className="format">
                  {quiz !== undefined &&
                    quiz.questions[questionNumber] !== undefined &&
                    quiz.questions[questionNumber].format}
                </p>
                {/* <p className="progress">
                  Question {questionNumber + 1}/{quiz.questions.length}
                </p> */}
              </div>
              <p className="question">{quiz.questions[questionNumber].title}</p>
            </header>
            {/* DISPLAYS THE CHOICES */}
            <main>
              <div className="choices">
                {quiz.questions[questionNumber].answers.map((answers, i) => (
                  <div
                    key={i}
                    className={this.handleSelection(
                      answers.answer.answer,
                      quiz.questions[questionNumber].title
                    )}
                    onClick={this.addAnswerToObject.bind(
                      this,
                      quiz.questions,
                      answers
                    )}
                  >
                    <p>
                      {answers.answer.answer}
                      {/* prints classnames of selected answer <span>
                        &nbsp;&nbsp;{this.handleSelection(
                          answers.answer.answer,
                          quiz.questions[questionNumber].title
                        )}
                      </span> */}
                    </p>

                    {/* <svg
                      height="100%"
                      width="100%"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <g>
                        <rect
                          className="rect-border"
                          height="100%"
                          width="100%"
                          fill="none"
                        />
                        <text x="20%" y="50" className="text">
                          {answers.answer.answer}
                        </text>
                      </g>
                    </svg> */}
                  </div>
                ))}
              </div>
              <div className="image-gallery-div">
                {quiz.questions[questionNumber].image !== undefined && (
                  <ImageGallery
                    items={[
                      {
                        original: quiz.questions[questionNumber].image,
                        thumbnail: quiz.questions[questionNumber].image
                      }
                    ]}
                    showThumbnails={false}
                    showPlayButton={false}
                    alt="question image"
                  />
                )}
              </div>
              <div className="buttons">
                {questionNumber !== 0 && (
                  <button
                    onClick={this.previousQuizQuestion(quiz.questions)}
                    className="previous-button"
                  >
                    previous
                  </button>
                )}
                {questionNumber + 1 !== quiz.questions.length ? (
                  <button
                    onClick={this.nextQuizQuestion(quiz.questions)}
                    className="next-button"
                  >
                    next
                  </button>
                ) : (
                  <button
                    onClick={this.submit(quiz.questions)}
                    className="submit-button"
                  >
                    Done
                  </button>
                )}
              </div>
            </main>
          </div>
        )}
      </div>
    )
  }
}

const mapStateToProps = (state, props) => ({
  username: state.auth.username,
  quiz:
    state.takeQuiz.quizAttemptInfoObj !== null &&
    state.takeQuiz.quizAttemptInfoObj,
  answers: state.takeQuiz.answers,
  answerArray: state.takeQuiz.answerArray,
  questionNumber: state.takeQuiz.questionNumber,
  multipleSelectAnswer: state.takeQuiz.multipleSelectAnswer,
  multipleChoiceAnswer: state.takeQuiz.multipleChoiceAnswer
})

export default connect(
  mapStateToProps,
  {
    startAddAnswerToArray,
    changeQuestionNumber,
    addMultipleChoiceAnswer,
    addMultipleSelectAnswer,
    updateMultipleSelectAnswer,
    updateAnswerArray,
    submitQuizAttempt,
    loadModal
  }
)(TakeQuizPage)
