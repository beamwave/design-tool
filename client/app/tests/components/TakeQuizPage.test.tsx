import React, { Component } from 'react'
import { shallow } from 'enzyme'
import { TakeQuizPage } from '../../components/TakeQuizPage'

test('should render TakeQuizPage component', () => {
  const wrapper = shallow(
    <TakeQuizPage
      username="Dilbert"
      quiz="placeholder quiz"
      questionNumber={18}
      multipleSelectAnswer="placeholder answer"
      multipleChoiceAnswer={['placeholder array of multiple choice']}
      answerArray={['placeholder array of answers']}
      loadModal={() => `placeholder function`}
      changeQuestionNumber={() => `placeholder function`}
      startAddAnswerToArray={() => `placeholder function`}
      addMultipleChoiceAnswer={() => `placeholder function`}
      addMultipleSelectAnswer={() => `placeholder function`}
      updateMultipleSelectAnswer={() => `placeholder function`}
      updateAnswerArray={() => `placeholder function`}
      submitQuizAttempt={() => `placeholder function`}
    />
  )
  expect(wrapper).toMatchSnapshot()
})
