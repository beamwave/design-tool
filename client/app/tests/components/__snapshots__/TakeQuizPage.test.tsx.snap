// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TakeQuizPage component 1`] = `
<div
  className="take-quiz-page"
>
  <Spinner
    className="loading-indicator"
    fadeIn="full"
    name="ball-spin-fade-loader"
    noFadeIn={false}
    overrideSpinnerClassName=""
  />
</div>
`;
